src/catalog/src/local/memory.rs

```diff
@@ -76,80 +76,75 @@ impl CatalogManager for MemoryCatalogManager {
     }
 
     async fn register_table(&self, request: RegisterTableRequest) -> Result<bool> {
-        let schema = self
-            .catalog(&request.catalog)
-            .context(CatalogNotFoundSnafu {
-                catalog_name: &request.catalog,
-            })?
-            .schema(&request.schema)
-            .await?
-            .context(SchemaNotFoundSnafu {
-                catalog: &request.catalog,
-                schema: &request.schema,
-            })?;
+        let catalog = request.catalog.clone();
+        let schema = request.schema.clone();
+        let result = self.register_table_sync(request);
         increment_gauge!(
             crate::metrics::METRIC_CATALOG_MANAGER_TABLE_COUNT,
             1.0,
-            &[crate::metrics::db_label(&request.catalog, &request.schema)],
+            &[crate::metrics::db_label(&catalog, &schema)],
         );
-        schema
-            .register_table(request.table_name, request.table)
-            .await
-            .map(|v| v.is_none())
+        result
     }
 
     async fn rename_table(&self, request: RenameTableRequest) -> Result<bool> {
-        let catalog = self
-            .catalog(&request.catalog)
-            .context(CatalogNotFoundSnafu {
+        let mut catalogs = self.catalogs.write().unwrap();
+        let schema = catalogs
+            .get_mut(&request.catalog)
+            .with_context(|| CatalogNotFoundSnafu {
                 catalog_name: &request.catalog,
+            })?
+            .get_mut(&request.schema)
+            .with_context(|| SchemaNotFoundSnafu {
+                catalog: &request.catalog,
+                schema: &request.schema,
             })?;
-        let schema =
-            catalog
-                .schema(&request.schema)
-                .await?
-                .with_context(|| SchemaNotFoundSnafu {
-                    catalog: &request.catalog,
-                    schema: &request.schema,
-                })?;
-        Ok(schema
-            .rename_table(&request.table_name, request.new_table_name)
-            .await
-            .is_ok())
+
+        // check old and new table names
+        if !schema.contains_key(&request.table_name) {
+            return TableNotFoundSnafu {
+                table_info: request.table_name.to_string(),
+            }
+            .fail()?;
+        }
+        if schema.contains_key(&request.new_table_name) {
+            return TableExistsSnafu {
+                table: &request.new_table_name,
+            }
+            .fail();
+        }
+
+        let table = schema.remove(&request.table_name).unwrap();
```

Comment:
Is it safe to unwrap here?

Reply:
Yes, this is checked before via `contains_key()`.